seedVault/index.ts

```diff
@@ -0,0 +1,131 @@
+export type StorageAdapter = {
+  get(key: string): Promise<string> | string | null;
+  set(key: string, value: string): Promise<void> | void;
+  remove(key: string): Promise<void> | void;
+};
+
+export type CryptoUtilsAdapter = {
+  encrypt(data: string, password: string): Promise<string>;
+  decrypt(data: string, password: string): Promise<string>;
+  hash(data: string, salt: string): Promise<string>;
+  generateRandomBytes(length: number): string | Promise<string>;
+};
+
+export type SeedVaultConfig = {
+  secureStorageAdapter: StorageAdapter;
+  cryptoUtilsAdapter: CryptoUtilsAdapter;
+  commonStorageAdapter: StorageAdapter;
+};
+
+export enum SeedVaultStorageKeys {
+  PASSWORD_HASH = 'passwordHash',
+  PASSWORD_SALT = 'passwordSalt',
+  ENCRYPTED_KEY = 'encryptedKey',
+  SEED_VAULT_VERSION = 'seedVaultVersion',
+}
+
+class SeedVault {
+  private readonly _secureStorageAdapter: StorageAdapter;
+
+  private readonly _cryptoUtilsAdapter: CryptoUtilsAdapter;
+
+  private readonly _commonStorageAdapter: StorageAdapter;
+
+  VERSION = 1;
+
+  constructor(config: SeedVaultConfig) {
+    this._secureStorageAdapter = config.secureStorageAdapter;
+    this._cryptoUtilsAdapter = config.cryptoUtilsAdapter;
+    this._commonStorageAdapter = config.commonStorageAdapter;
+  }
+
+  init = async (password: string) => {
+    this._commonStorageAdapter.set(SeedVaultStorageKeys.SEED_VAULT_VERSION, this.VERSION.toString());
+    const salt = await this._cryptoUtilsAdapter.generateRandomBytes(32);
+    if (!salt) throw new Error('Salt not set');
+    const passwordHash = await this._cryptoUtilsAdapter.hash(password, salt);
+    if (!passwordHash) throw new Error('Password hash not set');
+    this._commonStorageAdapter.set(SeedVaultStorageKeys.PASSWORD_SALT, salt);
+    this._secureStorageAdapter.set(SeedVaultStorageKeys.PASSWORD_HASH, passwordHash);
+  };
+
+  storeSeed = async (seed: string, overwriteExistingSeed?: boolean) => {
```

Comment:
I don't know if we should expose this like this 🤔 What about doing some thing like:

```
storeseed = async (seed: string) => this._storeseed(seed, false)

private _storeseed = async (seed: string, overwriteExistingSeed?: boolean) => {
...
```